### PR TITLE
Removing `run-tests.py` file.

### PR DESCRIPTION
**Pull request by prady0t**
# Description

Trying to remove run-tests.py file.

### CODE REVIEW

**Comment by prady0t** [comment]
Yet to add `conftest.py` file

File: noxfile.py
Comment (agriyakhetarpal) [inline noxfile.py:130]:
```suggestion
    session.run("python", "-m", "pytest", "-v", "tests/test_examples.py")
```

Soon enough, we should enable more options here when this PR is closer to merging. Personally, I like to use `pytest -svra`.

**Review by agriyakhetarpal — COMMENTED** [review]
This is looking good! I tried locally and

```bash
==================================================================== 44 passed in 100.46s (0:01:40) ====================================================================
nox > Session scripts was successful.
```

much faster than the serial implementation. I checked and the slowest scripts are `run_ecm.py`, `experiment_drive_cycle.py`,  and `pouch_cell_cooling.py`. It might be worth improving those, but I guess 100 seconds for 44 scripts isn't too bad at all, afterall, and is rather great in terms of speedups.

File: noxfile.py
Comment (kratman) [reply noxfile.py:130]: Do the options have any impact on the speed of the tests. I assume it would be negligible, but I am curious

Comment (arjxn-py) [reply noxfile.py:130]: I too believe that i'd be negligible since there's no extensive logging but I guess we can definitely do some benchmarking before merging to quantify the range of time difference

Comment (agriyakhetarpal) [reply noxfile.py:130]: Yes, as Arjun says, the impact is indeed negligible and it isn't measurable either; `-v` is to increase the verbosity, `-s` controls any print statements or logging statements in stdout, and `-ra` will display a summary of everything: failed + skipped + xfailed/xpassed + errors, after the test suite completes.

Comment (agriyakhetarpal) [reply noxfile.py:130]: For reference, after we'll have most of this ready and when we'll be starting with the migration of the tests (which is from this week onwards), a useful guide with some tips is present at https://learn.scientific-python.org/development/guides/pytest/

Comment (prady0t) [reply noxfile.py:129]: should i include `-svra` in all nox commands?

**Comment by prady0t** [comment]
I think I messed up the branch. I'm having problems while pushing changes. I've made the changes.

File: noxfile.py
Comment (agriyakhetarpal) [reply noxfile.py]: Yes, should be okay

Comment (prady0t) [reply noxfile.py]: Output with `-svra` is actually pretty clean. Thanks for the suggestion!

**Comment by kratman** [comment]
@prady0t Is your branch still having problems? Do you need help fixing the branch?

**Comment by prady0t** [comment]
> @prady0t Is your branch still having problems? Do you need help fixing the branch?

I can work with this, although I made a lot of commits.

**Comment by agriyakhetarpal** [comment]
No worries, we'll squash merge anyway, but to learn how to rewrite history would be up to you as an exercise

File: conftest.py
Comment (agriyakhetarpal) [inline conftest.py:8]:
```suggestion
        "--scripts", action="store_true", default=False, help="execute the example scripts"
```

File: conftest.py
Comment (agriyakhetarpal) [inline conftest.py:22]:
```suggestion
    config.addinivalue_line("markers", "scripts: mark test as an example script")
```

File: noxfile.py
Comment (agriyakhetarpal) [inline noxfile.py:107]: I know I suggested adding this earlier, but I think we have something better. Let's add these to `addopts` in `pyproject.toml` under `[tool.pytest.ini_options]` instead.
```suggestion
    session.run("python", "-m", "pytest", "-m", "unit")
```

Comment (agriyakhetarpal) [reply noxfile.py:107]: Also, can we make them run with `pytest --unit` instead of `pytest -m unit`? No big deal if we can't, just wanted to know if it's possible to do so with how you have configured the markers

File: noxfile.py
Comment (agriyakhetarpal) [inline noxfile.py:131]:
```suggestion
    session.run("python", "-m", "pytest", "--scripts", "tests/test_scripts.py")
```

Comment (agriyakhetarpal) [reply noxfile.py:131]: For example, this session runs `pytest --scripts`, while the `unit` session runs `pytest -m unit` (related to my question above). I think `pytest --unit`, `pytest --integration`, etc. are shorter and better to use.

Also, it would be great if we could configure `pytest --scripts` to run just `tests/test_scripts.py` itself, rather than us having to specify the file.

File: noxfile.py
Comment (agriyakhetarpal) [inline noxfile.py:162]:
```suggestion
    session.run("python", "-m", "pytest")
```

This command will run not only the unit and integration tests but also the example scripts, which contradicts the docstring for the function. Can we change it to `pytest --unit --integration` or similar?

File: pyproject.toml
Comment (agriyakhetarpal) [inline pyproject.toml:241]: If we were to add both `tests/` and `docs/source/examples/` here, could we skip specifying the path in `pytest --nbmake` above?

File: pyproject.toml
Comment (agriyakhetarpal) [inline pyproject.toml:223]:
```suggestion
minversion = "8"
```

Let us bump this to `8` here because `6` is too old now (please update in the `[dev]` extras and the installation guide table in the docs as well).

File: pyproject.toml
Comment (agriyakhetarpal) [inline pyproject.toml:231]: We can now add `-svra` to `addopts` below this line

File: tests/test_scripts.py
Comment (agriyakhetarpal) [inline tests/test_scripts.py:8]: We can use `pathlib.Path()` even further by replacing this as follows:

```suggestion
ROOT_DIR = Path(__file__).parent.parent
```

which is cleaner!

File: tests/test_scripts.py
Comment (agriyakhetarpal) [inline tests/test_scripts.py:19]:
```suggestion
    def list_of_files():
        file_list = (ROOT_DIR / "examples" / "scripts").rglob("*.py")
        return file_list
```

File: tests/testcase.py
Comment (agriyakhetarpal) [inline tests/testcase.py:9]:
```suggestion
    Custom TestCase class for PyBaMM
```

File: tests/testcase.py
Comment (agriyakhetarpal) [inline tests/testcase.py:13]: Note: could you move `assertDomainEqual` to `shared.py` in another PR later? (I think we discussed this somewhere)

Comment (agriyakhetarpal) [reply tests/testcase.py:13]: Never mind, you've already done it in https://github.com/pybamm-team/PyBaMM/pull/4231!

**Comment by agriyakhetarpal** [comment]
@prady0t, could you fix the failure in the style jobs so that the tests can run?

**Comment by prady0t** [comment]
> @prady0t, could you fix the failure in the style jobs so that the tests can run?

Done. I removed lot of code lines from `conftest.py` file and now we are using `-m` for everything. If a user runs pytest without any marker value, it runs all the tests, otherwise only those specified by `-m` flag.

File: noxfile.py
Comment (agriyakhetarpal) [reply noxfile.py]: Discussed on Slack, and `pytest -m unit` is the better one to use

File: pyproject.toml
Comment (agriyakhetarpal) [reply pyproject.toml:242]: Resolved in the meeting and we'll have to pass the path to the notebooks explicitly, it's already how it is like in our documentation: https://docs.pybamm.org/en/stable/source/user_guide/contributing.html#testing-the-example-notebooks

**Comment by prady0t** [comment]
Added documentation fixes.

File: CONTRIBUTING.md
Comment (agriyakhetarpal) [inline CONTRIBUTING.md:156]:
```suggestion
```

We should let go of the warning for now. Not many people ask about how to run tests, and I don't think this needs emphasis, because we are removing any mentions of `unittest` here anyway.

File: tests/test_scripts.py
Comment (agriyakhetarpal) [inline tests/test_scripts.py:22]: One thing I'm finding difficult with this approach right now is about seeing the names of the files – `pytest` displays just numbers, and the `-s` option means that it prints out a lot of information, coupled with the additional logging we've set through the fixture. Is it possible to maintain a clear separation between the file names and actually print the file names in the test summary info at the end? You may refer to the difference between the current logs and the logs in other PRs. I don't think the verbose logging is helping all that much – scrolling through 14K+ lines of logs isn't useful to us.

File: CONTRIBUTING.md
Comment (agriyakhetarpal) [inline CONTRIBUTING.md:154]: Could we revise this to say something on the lines: "we use the following plugins:" followed by a one-line description of the three of them?

File: tests/test_scripts.py
Comment (arjxn-py) [reply tests/test_scripts.py:22]: I agree with that, unit tests logs are also a little too much ~75,000.

About improving the clarity of test output, maybe we can try modifying the way pytest parametrize displays the file names.

Instead of just displaying numbers, we can try displaying a more descriptive name by using `pytest.param` while returning file list, try returning a `pytest.param` object with an id set to the file name like `[pytest.param(file, id=file.name) for file in file_list]`

**Review by arjxn-py — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md
Comment (agriyakhetarpal) [inline CONTRIBUTING.md:176]:
```suggestion
Next, add some simple (and speedy!) tests of your main features. If these run without exceptions that's a good start! Next, check the output of your methods using [assert statements](https://docs.pytest.org/en/7.1.x/how-to/assert.html).
```

File: pyproject.toml
Comment (agriyakhetarpal) [inline pyproject.toml:240]:
```suggestion
    "-vra",
```

I figured out the problem with the extra lines of logging output in the tests for the example scripts – this slipped right under our noses. This should be `-a`, not `-A` – the latter prints out the outputs from the tests that passed, and not just their names. Using `-a` should reduce the length of the logs by a significant amount.

**Review by agriyakhetarpal — APPROVED** [review]
Thanks for your work here, @prady0t! This looks great now – just a couple of comments, and this is ready to go.

File: tests/test_scripts.py
Comment (agriyakhetarpal) [reply tests/test_scripts.py]: This should be resolved by https://github.com/pybamm-team/PyBaMM/pull/4180#discussion_r1693373168.

**Comment by agriyakhetarpal** [comment]
This will need approval from @arjxn-py since he previously requested changes, we can merge after that

**Review by arjxn-py — APPROVED** [review]
This looks good, thanks @prady0t.
Will just wait for CI to pass to observe logs once, then we can merge.

**Comment by agriyakhetarpal** [comment]
No unnecessary logs this time around: https://github.com/pybamm-team/PyBaMM/actions/runs/10115401416/job/27976388229

**Comment by arjxn-py** [comment]
> No unnecessary logs this time around: https://github.com/pybamm-team/PyBaMM/actions/runs/10115401416/job/27976388229

Noticed that, just that we'd need to update branch to merge. My bad, I should have done it earlier.

**Pull request merged**
